### PR TITLE
`REVM`: 3.x Migration

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

REVM needed to be updated to the latest version (3.x) so we can start working on the big rewrites & new features.

## Solution

Migrates types/imports to conform to 3.x. These changes are focused on converting types—I'm deliberately choosing to keep the existing types that don't conflict there for now. The strategy is more or less as follows:

All of the types that need to use the revm types (e.g trait impls or structs/traits that interact very closely with revm) are switched.
Everything else is just converted when needed (see evm crate utils for utility functions)

Note that I've let _some_ lints purposely fail because these are areas that need confirmation as logic changed around them.

There's also the [old](https://github.com/mattsse/foundry/pull/2) PR for historical comments.

### CODE REVIEW

File: anvil/src/eth/backend/executor.rs [hunk 1]
```diff
@@ -258,11 +267,31 @@ impl<'a, 'b, DB: Db + ?Sized, Validator: TransactionValidator> Iterator
...
         // transact and commit the transaction
-        let ExecutionResult { exit_reason, out, gas_used, logs, .. } =
-            evm.inspect_commit(&mut inspector);
+        let exec_result = match evm.inspect_commit(&mut inspector) {
+            Ok(exec_result) => exec_result,
+            Err(err) => {
+                warn!(target: "backend", "[{:?}] failed to execute: {:?}", transaction.hash(), err);
+                return Some(TransactionExecutionOutcome::DatabaseError(
+                    transaction,
+                    DatabaseError::TransactionNotFound(H256::from_str("0x").unwrap()),
```
Comment (Evalir) [inline anvil/src/eth/backend/executor.rs:276]: Note that this error is very much a placeholder—We probably need another variant to properly represent the inspector failing.

File: anvil/src/eth/backend/executor.rs [hunk 2]
```diff
@@ -258,11 +267,31 @@ impl<'a, 'b, DB: Db + ?Sized, Validator: TransactionValidator> Iterator
...
-        if exit_reason == Return::OutOfGas {
+        let (exit_reason, gas_used, out, logs) = match exec_result {
+            ExecutionResult::Success { reason, gas_used, logs, output, .. } => {
+                (eval_to_instruction_result(reason), gas_used, Some(output), Some(logs))
+            }
+            ExecutionResult::Revert { gas_used, .. } => {
+                (InstructionResult::Revert, gas_used, None, None)
+            }
+            ExecutionResult::Halt { reason, gas_used } => {
+                (halt_to_instruction_result(reason), gas_used, None, None)
```
Comment (Evalir) [inline anvil/src/eth/backend/executor.rs:290]: Unsure if the gas refund calculation is accurate here (and on the other places this `ExecutionResult` is matched)

File: anvil/src/eth/backend/mem/mod.rs [hunk 3]
```diff
@@ -957,16 +989,30 @@ impl Backend {
...
     {
         let mut inspector = Inspector::default();
         let mut evm = revm::EVM::new();
         evm.env = self.build_call_env(request, fee_details, block_env);
         evm.database(state);
-        let (ExecutionResult { exit_reason, out, gas_used, .. }, state) =
-            evm.inspect_ref(&mut inspector);
+        let result_and_state = match evm.inspect_ref(&mut inspector) {
+            Ok(result_and_state) => result_and_state,
+            Err(e) => return Err(BlockchainError::EvmError(InstructionResult::FatalExternalError)),
```
Comment (Evalir) [inline anvil/src/eth/backend/mem/mod.rs:1002]: Also using this error as a placeholder—probably also need another variant for this type of inspector error

**Review by mattsse — CHANGES_REQUESTED** [review]
incredible work!

even if all the conversions are a bit horrible, we need to do this to support revm and they will become obsolete eventually.

could you please enable contributions by mainter? I have a few local changes that I'd like to push.
I think the issue is that you opened with master branch which is perhaps protected?

a few tests hang but I suspect they're are related to infura/goerli rpc keys.
so before we continue we need to update those, ref https://github.com/foundry-rs/foundry/pull/4781

but this looks pretty good, can merge once tests green!

**Comment by Evalir** [comment]
Aight all tests seem to be fixed but CI is hanged on this last one (`test_cmd::can_test_forge_std`). As you mentioned @mattsse it could probably be the keys—how should we proceed?

**Review by mattsse — APPROVED** [review]
alright, let's do a final rebase then send it

**Comment by sambacha** [comment]
I christen thee, v1.0.0

**Pull request merged**